client.go

```diff
@@ -91,167 +72,88 @@ func (c *Client) Dial(address string) (conn *Conn, err error) {
 }
 
 // NewMessage Create message for request
-func (co *Conn) NewMessage(p MessageParams) Message {
-	switch co.Conn.(type) {
-	case *net.TCPConn, *tls.Conn:
-		return NewTcpMessage(p)
-	}
-	return NewDgramMessage(p)
-}
-
-// ReadMsg reads a message from the connection co.
-func (co *Conn) ReadMsg(timeout time.Duration) (Message, error) {
-	var m []byte
-	var err error
-
-	co.SetReadDeadline(time.Now().Add(timeout))
-	for {
-		m, err = co.Read()
-		if err == ErrShortRead {
-			continue
-		} else if err != nil {
-			return nil, err
-		}
-		break
-	}
-	switch co.Conn.(type) {
-	case *net.TCPConn, *tls.Conn:
-		resp, _, err := PullTcp(m)
-		return resp, err
-	}
-	// UDP connection
-	return ParseDgramMessage(m)
-}
-
-func tcpReadBuf(conn *Conn) ([]byte, error) {
-	r := bytes.NewReader(conn.TCPbuf)
-	mti, err := readTcpMsgInfo(r)
-	if err != nil {
-		return nil, ErrShortRead
-	}
-	if len(conn.TCPbuf) < mti.totLen {
-		return nil, ErrShortRead
-	}
-	buf := conn.TCPbuf[:mti.totLen]
-	conn.TCPbuf = conn.TCPbuf[mti.totLen:]
-	return buf, nil
-}
-
-// tcpRead calls TCPConn.Read enough times to fill allocated buffer.
-func tcpRead(conn *Conn) ([]byte, error) {
-	m := make([]byte, maxPktLen)
-	n, err := conn.Conn.Read(m)
-	if err != nil {
-		return nil, err
-	}
-	m = m[:n]
-	conn.TCPbuf = append(conn.TCPbuf, m...)
-	return tcpReadBuf(conn)
-}
-
-// Read implements the net.Conn read method.
-func (co *Conn) Read() ([]byte, error) {
-	switch co.Conn.(type) {
-	case *net.TCPConn, *tls.Conn:
-		m, err := tcpReadBuf(co)
-		if err == nil {
-			return m, nil
-		}
-		return tcpRead(co)
-	}
-	// UDP connection
-	m := make([]byte, maxPktLen)
-	n, err := co.Conn.Read(m)
-	if err != nil {
-		return nil, err
-	}
-	return m[:n], err
+func (co *ClientConn) NewMessage(p MessageParams) Message {
+	return co.session.NewMessage(p)
 }
 
 // WriteMsg sends a message through the connection co.
-func (co *Conn) WriteMsg(m Message, timeout time.Duration) (err error) {
-	var data []byte
-	data, err = m.MarshalBinary()
-	if err != nil {
-		return err
-	}
-	co.SetWriteDeadline(time.Now().Add(timeout))
-	_, err = co.Write(data)
-	return err
-}
-
-// Write implements the net.Conn Write method.
-func (co *Conn) Write(p []byte) (n int, err error) {
-	return co.Conn.Write(p)
+func (co *ClientConn) WriteMsg(m Message) (err error) {
+	return co.session.WriteMsg(m)
 }
 
-// Return the appropriate timeout for a specific request
-func (c *Client) getTimeoutForRequest(timeout time.Duration) time.Duration {
-	var requestTimeout time.Duration
-	if c.Timeout != 0 {
-		requestTimeout = c.Timeout
-	} else {
-		requestTimeout = timeout
-	}
-	// net.Dialer.Timeout has priority if smaller than the timeouts computed so
-	// far
-	if c.Dialer != nil && c.Dialer.Timeout != 0 && c.Dialer.Timeout < requestTimeout {
-		requestTimeout = c.Dialer.Timeout
-	}
-	return requestTimeout
+// Close close connection
+func (co *ClientConn) Close() {
+	co.srv.Shutdown()
+	<-co.shutdownSync
```

Comment:
same timeout pattern, what if co fails to send? this will hang. 
with timeout switch you can give it some time to shutdown and if it fails just kill it

Reply:
Fixed